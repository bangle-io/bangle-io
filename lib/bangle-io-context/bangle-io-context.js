import { SpecRegistry } from '@bangle.dev/core/spec-registry';

function filterFlatMap(array, field, flatten = true) {
  let items = array.filter((item) => Boolean(item[field]));
  if (flatten) {
    return items.flatMap((item) => item[field]);
  }
  return items.map((item) => item[field]);
}

export class BangleIOContext {
  constructor({
    coreRawSpecs,
    getCorePlugins,
    extensions = [],
    markdownItPlugins = [],
  }) {
    this._extensions = extensions;
    if (
      new Set(
        this._extensions.filter((r) => Boolean(r.name)).map((r) => r.name),
      ).size !== this._extensions.length
    ) {
      throw new Error('Extension name must be unique');
    }
    this._getCorePlugins = getCorePlugins;
    this.specRegistry = new SpecRegistry([
      ...coreRawSpecs,
      ...filterFlatMap(extensions, 'editorSpecs'),
    ]);
    this.markdownItPlugins = [
      ...markdownItPlugins,
      ...filterFlatMap(extensions, 'markdownItPlugins'),
    ];

    this._renderReactNodeViewLookup = Object.fromEntries(
      filterFlatMap(extensions, 'renderReactNodeView', false).flatMap((obj) =>
        Object.entries(obj),
      ),
    );
  }

  _getExtensionStore(uniqueObj, extension) {
    if (!uniqueObj[extension.name]) {
      uniqueObj[extension.name] = {};
    }
    return uniqueObj[extension.name];
  }

  renderReactNodeViews({ nodeViewRenderArg, wsPath, editorId }) {
    return this._renderReactNodeViewLookup[nodeViewRenderArg.node.type.name]?.({
      nodeViewRenderArg,
      wsPath,
      editorId,
      bangleIOContext: this,
    });
  }

  getPlugins() {
    return [
      ...filterFlatMap(this._extensions, 'highPriorityEditorPlugins'),
      this._getCorePlugins(),
      ...filterFlatMap(this._extensions, 'editorPlugins'),
    ];
  }

  editor = {
    beforeDestroy: ({ uniqueEditorObj, wsPath, editorId }) => {
      for (const ext of this._extensions.filter(
        (e) => e.editor?.beforeDestroy,
      )) {
        ext.editor.beforeDestroy({
          wsPath,
          editorId,
          store: this._getExtensionStore(uniqueEditorObj, ext),
        });
      }
    },
    initialScrollPos: ({
      uniqueEditorObj,
      wsPath,
      editorId,
      scrollParent,
      doc,
    }) => {
      for (const ext of this._extensions) {
        const result = ext.editor?.initialScrollPos?.({
          wsPath,
          editorId,
          scrollParent,
          doc,
          store: this._getExtensionStore(uniqueEditorObj, ext),
        });
        if (result != null) {
          return result;
        }
      }

      return undefined;
    },

    initialSelection: ({ uniqueEditorObj, wsPath, editorId, doc }) => {
      for (const ext of this._extensions) {
        const result = ext.editor?.initialSelection?.({
          wsPath,
          editorId,
          doc,
          store: this._getExtensionStore(uniqueEditorObj, ext),
        });
        if (result != null) {
          return result;
        }
      }
      return undefined;
    },
  };
}
