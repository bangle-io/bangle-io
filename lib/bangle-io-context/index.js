export * from './bangle-io-context';
