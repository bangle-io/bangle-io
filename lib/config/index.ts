import { compileConfig } from './compile-config';

export {
  browserInfo,
  isChrome,
  isFirefox,
  isMac,
  isMobile,
  isSafari,
} from './browser';
export const config = compileConfig();
export type { FinalConfig } from '@bangle.io/config-template';

// appEnv can be one of the following only `production`, `staging`,
// `local` , `dev/*` where * is the branch name
export const APP_ENV: string = config.build.appEnv;
export const IS_PRODUCTION_APP_ENV = APP_ENV === 'production';
export const RELEASE_VERSION: string = config.build.releaseVersion;
export const RELEASE_ID: string = config.build.releaseId;
// a less intimidating thing that is shown in the UI
// for production it is release version but for other env we show the whole thing
// for better debugging
export const FRIENDLY_ID = IS_PRODUCTION_APP_ENV ? RELEASE_VERSION : RELEASE_ID;

export const HELP_DOCS_VERSION: string = config.app.helpDocsVersion;
export const TAB_ID: string = 'tab_' + randomStr(4);
export const BANGLE_HOT = config.build.hot;
export const CHANGELOG_TEXT: string = config.app.changelogText;
export const IS_STORYBOOK = config.build.storybook;

export const DEBUG_WRITE_SLOWDOWN = config.debug?.writeSlowDown;
export const IS_TEST_ENV = config.build.nodeEnv === 'test';

export const sentryConfig = {
  environment: APP_ENV,
  dsn: 'https://f1a3d53e530e465e8f74f847370b594b@o573373.ingest.sentry.io/5723848',
  integrations: [],
  release: RELEASE_ID,
  tracesSampleRate:
    APP_ENV === 'production' ? 0.8 : APP_ENV === 'staging' ? 1 : 0,
};

if (config.build.nodeEnv !== 'test') {
  console.log(config.build.appEnv + ': using ' + RELEASE_ID);

  console.table({
    tabId: TAB_ID,
    isWorkerContext:
      typeof WorkerGlobalScope !== 'undefined' &&
      // eslint-disable-next-line no-restricted-globals, no-undef
      self instanceof WorkerGlobalScope,
  });
}

if (!/^\d+\.\d+\.\d+/.test(HELP_DOCS_VERSION || '')) {
  throw new Error('Invalid HELP_DOCS_VERSION: ' + HELP_DOCS_VERSION);
}

// WARNING: the width is hard coded at multiple places, search for it
// by value if you want to change it
export const SPLIT_SCREEN_MIN_WIDTH = 759;
export const FILE_PALETTE_MAX_RECENT_FILES = 15;
export const FILE_PALETTE_MAX_FILES = 200;
export const SERVICE_WORKER_UPDATE_INTERVAL = IS_PRODUCTION_APP_ENV
  ? 10 * 60 * 1000
  : 20 * 1000;

function randomStr(len = 10) {
  return Math.random().toString(36).substring(2, 15).slice(0, len);
}
