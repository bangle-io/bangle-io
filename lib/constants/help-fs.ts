import type { WorkspaceInfo } from '@bangle.io/shared-types';

import { WorkspaceTypeHelp } from './workspace';

export const HELP_FS_WORKSPACE_NAME = 'bangle-help';
export const BANGLE_HOME_PATH = `/ws/${WorkspaceTypeHelp}`;
export const HELP_FS_INDEX_FILE_NAME = 'getting started.md';
export const HELP_FS_INDEX_WS_PATH = `${HELP_FS_WORKSPACE_NAME}:${HELP_FS_INDEX_FILE_NAME}`;

let cachedHelpFs: WorkspaceInfo | undefined = undefined;

export const helpFSWorkspaceInfo = (): WorkspaceInfo => {
  if (!cachedHelpFs) {
    cachedHelpFs = {
      metadata: {
        allowLocalChanges: true,
      },
      name: HELP_FS_WORKSPACE_NAME,
      type: WorkspaceTypeHelp,
      lastModified: Date.now(),
    };
  }

  return cachedHelpFs;
};
