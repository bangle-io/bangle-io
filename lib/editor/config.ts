export const EDITOR_WATCH_PLUGIN_HOST_WAIT_TIME = 30;
