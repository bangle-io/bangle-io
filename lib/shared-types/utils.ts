export type FirstParameter<T extends (...args: any) => any> = Parameters<T>[0];
