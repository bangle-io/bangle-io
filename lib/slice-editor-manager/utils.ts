import type { BangleEditor } from '@bangle.dev/core';
import type { EditorView } from '@bangle.dev/pm';

import { MAX_OPEN_EDITORS } from '@bangle.io/constants';
import { getEditorPluginMetadata } from '@bangle.io/editor-common';
import { z } from '@bangle.io/nsm-3';
import type { EditorIdType, WsPath } from '@bangle.io/shared-types';
import { findWrappingScrollable } from '@bangle.io/utils';

import type { EditorSliceState } from './types';

// In PM selection.toJSON() is of different types based on the selection type
// we only care about serializing and this type is good enough.
export type SelectionJson = z.infer<typeof selectionJsonSchema>;
// WARNING: before changing this schema, make sure to update the
//  make sure to update the manual check in `calculateSelection`
export const selectionJsonSchema = z.record(
  z.union([z.undefined(), z.string(), z.number(), z.null()]),
);

export const calculateSelection = (
  editorId: EditorIdType,
  editor: BangleEditor,
): {
  wsPath: WsPath;
  editorId: EditorIdType;
  selectionJson: SelectionJson;
} => {
  const selection = editor.view.state.selection;

  const json = selection.toJSON();

  for (const value of Object.values(json)) {
    if (
      typeof value !== 'number' &&
      typeof value !== 'string' &&
      value != null
    ) {
      throw new Error(`Invalid editor selection json value type: ${json}`);
    }
  }

  return {
    wsPath: getEditorPluginMetadata(editor.view.state).wsPath,
    editorId: editorId,
    selectionJson: json as SelectionJson,
  };
};

export const calculateScrollPosition = (
  editorId: EditorIdType,
  view: EditorView,
) => {
  const top = getScrollParentElement(editorId)?.scrollTop;

  if (typeof top === 'number') {
    return {
      wsPath: getEditorPluginMetadata(view.state).wsPath,
      editorId: editorId,
      scrollPosition: top,
    };
  }

  return undefined;
};

export function* getEachEditorIterable(
  editorManagerState: EditorSliceState,
): Iterable<{
  editor: BangleEditor | undefined;
  editorId: NonNullable<EditorIdType>;
}> {
  for (const [index, editor] of editorManagerState.mainEditors.entries()) {
    yield { editor, editorId: index };
  }
}

export function isValidEditorId(editorId: EditorIdType): boolean {
  return editorId >= 0 && editorId < MAX_OPEN_EDITORS;
}

export function assertValidEditorId(
  editorId: EditorIdType,
): asserts editorId is EditorIdType {
  if (!isValidEditorId(editorId)) {
    throw new Error('editorId is out of range or invalid');
  }
}

export function getScrollParentElement(editorId: EditorIdType) {
  const editor = document.querySelector(
    '.B-editor-container_editor-' + editorId,
  );

  if (editor) {
    return findWrappingScrollable(editor);
  }

  return undefined;
}
