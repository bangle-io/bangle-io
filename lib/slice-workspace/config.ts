import { isMobile } from '@bangle.io/utils';

export const MAX_ENTRIES = isMobile ? 12 : 64;
export const MAX_TIMESTAMPS_PER_ENTRY = 5;
