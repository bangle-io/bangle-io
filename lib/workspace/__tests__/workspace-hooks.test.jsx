import mockBabyFs from './baby-fs.mock';

import { render, act } from '@testing-library/react';
import React from 'react';
import { MemoryRouter as Router, Switch, Route } from 'react-router-dom';
import * as idb from 'idb-keyval';
import { checkWidescreen } from 'utils/index';
import { BangleIOContext } from 'bangle-io-context/index';
import { defaultSpecs } from '@bangle.dev/core/test-helpers/default-components';
import { Workspace } from '../Workspace';
import {
  useCreateNote,
  useDeleteFile,
  useListCachedNoteWsPaths,
  useWorkspacePath,
  useWorkspaces,
} from '../workspace-hooks';
import { helpFSWorkspaceInfo } from 'config/help-fs';

jest.mock('utils/index', () => {
  const actual = jest.requireActual('utils/index');

  return {
    ...actual,
    checkWidescreen: jest.fn(() => false),
  };
});

function createFileContent(textContent = 'hello') {
  return JSON.stringify({
    content: [
      {
        content: [{ text: textContent, type: 'text' }],
        type: 'paragraph',
      },
    ],
    type: 'doc',
  });
}

const originalFile = window.File;

beforeEach(() => {
  window.File = class File {
    constructor(content, fileName, opts) {
      this.content = content;
      this.fileName = fileName;
      this.opts = opts;
    }
  };
});

afterEach(() => {
  window.File = originalFile;
});

describe('useListCachedNoteWsPaths', () => {
  const App = ({ Comp }) => (
    <Router initialEntries={['/ws/kujo']}>
      <Switch>
        <Route path={['/ws/:wsName']}>
          <Workspace>
            <Comp />
          </Workspace>
        </Route>
      </Switch>
      <Route
        path="*"
        render={({ history, location }) => {
          return null;
        }}
      />
    </Router>
  );

  test('works', async () => {
    let refreshFiles;

    mockBabyFs.setupMockWorkspace({ name: 'kujo' });

    await mockBabyFs.setupMockFile('kujo', 'one.md');

    function Comp() {
      const [files, _refreshFiles] = useListCachedNoteWsPaths();
      refreshFiles = _refreshFiles;
      return (
        <div data-testid="result">
          {files && files.map((f) => <span key={f}>{f}</span>)}
        </div>
      );
    }

    let promise = Promise.resolve();
    let result;
    act(() => {
      result = render(<App Comp={Comp} />);
    });
    // To let the initial setState settle in Workspace after mount
    await act(() => promise);

    expect(result.container).toMatchInlineSnapshot(`
      <div>
        <div
          data-testid="result"
        >
          <span>
            kujo:one.md
          </span>
        </div>
      </div>
    `);

    await mockBabyFs.idbFS.writeFile('kujo/two.md', createFileContent());
    await mockBabyFs.setupMockFile('kujo', 'two.md');

    await act(async () => {
      await refreshFiles();
    });

    expect(result.container).toMatchInlineSnapshot(`
      <div>
        <div
          data-testid="result"
        >
          <span>
            kujo:one.md
          </span>
          <span>
            kujo:two.md
          </span>
        </div>
      </div>
    `);
  });

  test('only returns md files', async () => {
    mockBabyFs.setupMockWorkspace({ name: 'kujo' });
    await mockBabyFs.setupMockFile('kujo', 'one.md');
    await mockBabyFs.setupMockFile('kujo', 'two.png');
    await mockBabyFs.setupMockFile('kujo', 'three.md');

    function Comp() {
      const [files] = useListCachedNoteWsPaths();
      return (
        <div data-testid="result">
          {files && files.map((f) => <span key={f}>{f}</span>)}
        </div>
      );
    }

    let promise = Promise.resolve();
    let result;
    act(() => {
      result = render(<App Comp={Comp} />);
    });
    // To let the initial setState settle in Workspace after mount
    await act(() => promise);

    expect(result.container).toMatchInlineSnapshot(`
      <div>
        <div
          data-testid="result"
        >
          <span>
            kujo:one.md
          </span>
          <span>
            kujo:three.md
          </span>
        </div>
      </div>
    `);
  });
});

describe('useWorkspacePath', () => {
  let testLocation;
  const App = ({ Comp }) => (
    <Router initialEntries={['/ws/kujo']}>
      <Switch>
        <Route path={['/ws/:wsName']}>
          <Workspace>
            <Comp />
          </Workspace>
        </Route>
      </Switch>
      <Route
        path="*"
        render={({ history, location }) => {
          testLocation = location;

          return null;
        }}
      />
    </Router>
  );

  beforeEach(() => {
    testLocation = null;
  });

  test('pushWsPath works', async () => {
    // to get secondary tested
    checkWidescreen.mockImplementation(() => true);

    let workspacePathHookResult;
    mockBabyFs.setupMockWorkspace({ name: 'kujo' });

    await mockBabyFs.setupMockFile('kujo', 'one.md');

    function Comp() {
      const [files] = useListCachedNoteWsPaths();
      workspacePathHookResult = useWorkspacePath();
      return (
        <div data-testid="result">
          {files && files.map((f) => <span key={f}>{f}</span>)}
        </div>
      );
    }

    let promise = Promise.resolve();
    let result;
    act(() => {
      result = render(<App Comp={Comp} />);
    });
    expect(workspacePathHookResult).toMatchSnapshot();

    // To let the initial setState settle in Workspace after mount
    await act(() => promise);

    expect(
      result.container.querySelectorAll('[data-testid="result"]'),
    ).toHaveLength(1);

    await act(async () => {
      await workspacePathHookResult.pushWsPath('kujo:two.md');
    });

    expect(testLocation?.pathname).toBe('/ws/kujo/two.md');
    expect(testLocation?.state).toMatchInlineSnapshot(`
      Object {
        "workspaceInfo": Object {
          "metadata": Object {},
          "name": "kujo",
          "type": "browser",
        },
        "workspaceStatus": "READY",
      }
    `);
    expect(workspacePathHookResult).toMatchSnapshot();

    await act(async () => {
      await workspacePathHookResult.removeWsPath();
    });
    expect(testLocation?.pathname).toBe('/ws/kujo');

    await act(async () => {
      await workspacePathHookResult.replaceWsPath('kujo:one.md');
    });
    expect(testLocation?.pathname).toBe('/ws/kujo/one.md');

    expect(testLocation.search).toBe('');

    await act(async () => {
      await workspacePathHookResult.pushWsPath(
        'kujo:three.md',
        undefined,
        true,
      );
    });
    expect(testLocation?.search).toBe('?secondary=kujo%3Athree.md');
    expect(testLocation?.pathname).toBe('/ws/kujo/one.md');

    await act(async () => {
      await workspacePathHookResult.removeSecondaryWsPath();
    });
    expect(testLocation?.search).toBe('');
    expect(testLocation?.pathname).toBe('/ws/kujo/one.md');
  });
});

describe('useCreateNote', () => {
  test('browser create file', async () => {
    let callback;
    let testLocation;

    mockBabyFs.setupMockWorkspace({ name: 'kujo' });

    function Comp() {
      const createMdFile = useCreateNote();
      callback = createMdFile;
      return <div>Hello</div>;
    }

    await render(
      <Router initialEntries={['/ws/kujo']}>
        <Switch>
          <Route path={['/ws/:wsName']}>
            <Comp />
          </Route>
        </Switch>
        <Route
          path="*"
          render={({ history, location }) => {
            testLocation = location;
            return null;
          }}
        />
      </Router>,
    );

    expect(callback).not.toBeUndefined();

    const result = callback(
      new BangleIOContext({
        coreRawSpecs: defaultSpecs(),
        markdownItPlugins: [],
      }),
      'kujo:one.md',
    );
    expect(result).toBeInstanceOf(Promise);

    await result;

    expect(testLocation.pathname).toBe('/ws/kujo/one.md');

    expect(mockBabyFs.idbFS.writeFile).toBeCalledTimes(1);
    expect(mockBabyFs.idbFS.writeFile).toBeCalledWith(
      'kujo/one.md',
      new File(['# one\n\nHello world!'], 'one.md', { type: 'text/plain' }),
    );
  });
});

describe('useDeleteFile', () => {
  test('browser delete file', async () => {
    let callback;
    let testLocation;

    mockBabyFs.setupMockWorkspace({ name: 'kujo' });

    await mockBabyFs.setupMockFile('kujo', 'one.md');

    function Comp() {
      const deleteFileCb = useDeleteFile();
      callback = deleteFileCb;
      return <div>Hello</div>;
    }

    await render(
      <Router initialEntries={['/ws/kujo/one.md']}>
        <Switch>
          <Route path={['/ws/:wsName']}>
            <Comp />
          </Route>
        </Switch>
        <Route
          path="*"
          render={({ history, location }) => {
            testLocation = location;
            return null;
          }}
        />
      </Router>,
    );

    expect(callback).not.toBeUndefined();

    const result = callback('kujo:one.md');
    expect(result).toBeInstanceOf(Promise);

    await result;

    expect(testLocation.pathname).toBe('/ws/kujo');

    expect(mockBabyFs.idbFS.unlink).toBeCalledTimes(1);
    expect(mockBabyFs.idbFS.unlink).toBeCalledWith('kujo/one.md');
  });
});

describe('useWorkspaces', () => {
  test('loads workspace on mount', async () => {
    mockBabyFs.mockStore.clear();
    let createWorkspace, testLocation;
    mockBabyFs.setupMockWorkspace({ name: 'kujo1' });

    function CompCreateWS() {
      const { workspaces = [] } = useWorkspaces();
      return <div data-testid="result">{workspaces.map((r) => r.name)}</div>;
    }

    let result;
    act(() => {
      result = render(
        <Router initialEntries={['/ws']}>
          <Switch>
            <Route path="/ws">
              <CompCreateWS />
            </Route>
            <Route exact path="/ws/:wsName">
              <CompCreateWS />
            </Route>
          </Switch>
          <Route
            path="*"
            render={({ history, location }) => {
              testLocation = location;
              return null;
            }}
          />
        </Router>,
      );
    });

    let promise = Promise.resolve();

    await act(() => promise);
    expect(result.container).toMatchInlineSnapshot(`
      <div>
        <div
          data-testid="result"
        >
          kujo1
          bangle-help
        </div>
      </div>
    `);
  });

  test('createWorkspace', async () => {
    mockBabyFs.mockStore.clear();
    let createWorkspace, testLocation;

    function CompCreateWS() {
      const r = useWorkspaces();
      createWorkspace = r.createWorkspace;
      return <div>Hello</div>;
    }

    act(() => {
      render(
        <Router initialEntries={['/ws']}>
          <Switch>
            <Route path="/ws">
              <CompCreateWS />
            </Route>
            <Route exact path="/ws/:wsName">
              <CompCreateWS />
            </Route>
          </Switch>
          <Route
            path="*"
            render={({ history, location }) => {
              testLocation = location;
              return null;
            }}
          />
        </Router>,
      );
    });

    await act(() => createWorkspace('kujo1'));

    expect(testLocation.pathname).toBe('/ws/kujo1');
    // Note: for some reason MemoryRouter doesnt do urlParams
    // correctly
    expect(idb.set).toBeCalledWith('workspaces/2', [
      helpFSWorkspaceInfo,
      { metadata: {}, name: 'kujo1', type: 'browser' },
    ]);
  });
});
