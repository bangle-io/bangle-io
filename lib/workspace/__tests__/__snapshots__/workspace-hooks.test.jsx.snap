// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWorkspacePath pushWsPath works 1`] = `
Object {
  "filePath": "",
  "pushWsPath": [Function],
  "refreshHistoryStateKey": [Function],
  "removeSecondaryWsPath": [Function],
  "removeWsPath": [Function],
  "replaceWsPath": [Function],
  "secondaryWsPath": undefined,
  "wsName": "kujo",
  "wsPath": undefined,
}
`;

exports[`useWorkspacePath pushWsPath works 3`] = `
Object {
  "filePath": "two.md",
  "pushWsPath": [Function],
  "refreshHistoryStateKey": [Function],
  "removeSecondaryWsPath": [Function],
  "removeWsPath": [Function],
  "replaceWsPath": [Function],
  "secondaryWsPath": undefined,
  "wsName": "kujo",
  "wsPath": "kujo:two.md",
}
`;
